SisoVectorSystem => VectorSystem

RussTedrake opened the pull request:
As discussed f2f w/ @jwnimmer-tri .

"SISO" was the wrong name here; that would have implied scalar input / output (not single vector input/output).  Also, I wanted to support zero inputs/output with the VectorSystem abstraction; as they are common and useful. 

Updated the system class "hello world" simple examples to use VectorSystem.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6501)
<!-- Reviewable:end -->

RussTedrake commented:
+@jwnimmer-tri for feature review.

---

Review status: 0 of 22 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-:-Ko99cSI2ZgDx5YM34OY:b-1tks9q)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.

---

Reviewed 4 of 25 files at r1.
Review status: 2 of 22 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/vector_system.h, line 52 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHnn_tQcMncGrqwXMz:-KoHnn_tQcMncGrqwXN-:b-ug2ioy) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L52)):*
> ```objc
>   void get_output_port(int) = delete;
> 
>   // Confirms the SISO invariants when allocating the context.
> ```

This use of "SISO" is now confusing, because the term is no longer defined above.

---

*[drake/systems/framework/vector_system.h, line 104 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHoP2_2LwWzJnpHfZ_:-KoHoP2_2LwWzJnpHfZa:bdxwz3) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L104)):*
> ```objc
>         (this->get_num_input_ports() > 0)
>             ? this->EvalEigenVectorInput(context, 0)
>             : Eigen::VectorBlock<const VectorX<T>>(VectorX<T>(0), 0, 0);
> ```

The lifetime of the temporary `VectorX<T>(0)` does not extend past this statement, but `input_block` continues to alias (and use) it.  (It should turn into a named local variable.)

Ditto for the identical code below.

---

*[drake/systems/framework/vector_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHowjUeIJXIyY58l7z:-KoHowjVXZ5raHgCqWcv:b9ly61v) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L164)):*
> ```objc
>   void CalcVectorOutput(const Context<T>& context,
>                         BasicVector<T>* output) const {
>     // Short-circuit when there's no work to do.
> ```

This method will only ever be called when there was an output port declared; thus, this guard is misleading.  (The best fix is probably to turn this into an assertion.)

Also, the vector size doesn't seems like the correct thing to check -- its the port count that matters, not the port width, I think?

---

*[drake/systems/framework/vector_system.h, line 173 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHpsB1DtbylkNa3CjP:-KoHpsB1DtbylkNa3CjQ:bbjxd4u) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L173)):*
> ```objc
>         (this->get_num_input_ports() > 0)
>             ? this->EvalEigenVectorInput(context, 0)
>             : Eigen::VectorBlock<const VectorX<T>>(VectorX<T>(0), 0, 0);
> ```

FYI This is the third time this (tricky) bit of code appears.  Normally that would be a defect, but in this case the purpose of the class is to deal with this boilerplate, and its not immediately clear that a factoring would be simpler.  But in addressing the defect (above), please consider whether factoring out this code could be better.

---

*[drake/systems/framework/vector_system.h, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsQMyVOfVKjwht4zW:-KoHsQMzRwhVyCx2v9bS:b-78qby5) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L206)):*
> ```objc
>   /// time.
>   ///
>   /// By default, this function does nothing if the @p derivatives are empty,
> ```

`@p derivatives` seems like a paste-o.

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsyf7En9QyfppQKN7:-KoHsyf7En9QyfppQKN8:bdjhoks) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
> ```objc
>       Eigen::VectorBlock<VectorX<T>>* output) const {
>     unused(context, input, state);
>     DRAKE_THROW_UNLESS(output->size() == 0);
> ```

Ditto on "the vector size doesn't seems like the correct thing to check -- its the port count that matters, not the port width, I think?".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-:-KoHt3TbzPIpQL6Mhjkm:by96gdd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

---

Reviewed 21 of 25 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[drake/BUILD, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1CSgcXR1IaUMqAhJ:-KoI1CSgcXR1IaUMqAhK:brnjynm) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/BUILD#L323)):*
> ```Python
>     guess_hdrs = "WORKSPACE",
>     hdr_dest = "include/drake",
>     targets = ["libdrake.so"],
> ```

The prior sort-order was correct.

---

*[drake/examples/simple_continuous_time_system.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHuJcEM-pwFqz_tsnB:-KoHuJcEM-pwFqz_tsnC:b7r9xgf) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_continuous_time_system.cc#L18)):*
> ```cpp
>   SimpleContinuousTimeSystem()
>       : drake::systems::VectorSystem<double>(0,
>                                              1) {  // Zero inputs, one output.
> ```

FYI Consider a comment on each line:
```
(0,    // Zero inputs.
 1) {  // One output.
```

---

*[drake/examples/simple_continuous_time_system.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI0jA5kv7xzts7Vt_-:-KoI0jA5kv7xzts7Vt_0:b-91ozwc) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_continuous_time_system.cc#L28)):*
> ```cpp
>       const Eigen::VectorBlock<const Eigen::VectorXd>& u,
>       const Eigen::VectorBlock<const Eigen::VectorXd>& x,
>       Eigen::VectorBlock<Eigen::VectorXd>* xdot) const {
> ```

FYI _Meta_ This override does not use the superclass's names for the parameters; usually this is a defect, though perhaps there is some reason why they should differ in this case?  Should the base class be changed to use these short names instead?

---

*[drake/examples/simple_discrete_time_system.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI0SSHD07BQxZTwlOH:-KoI0SSHD07BQxZTwlOI:bbhcekg) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_discrete_time_system.cc#L28)):*
> ```cpp
>       const Eigen::VectorBlock<const Eigen::VectorXd>& u,
>       const Eigen::VectorBlock<const Eigen::VectorXd>& x,
>       Eigen::VectorBlock<Eigen::VectorXd>* xn) const {
> ```

... and in particular (re: renaming base class's paramter)s , `xn` seems like a really bad name, since this is `x[n+1]` not `x[n]` (which is named `x` above).

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI7Laq_11NuoKiMC6C:-KoI7Lar9i28V2yF_6EI:b437fjg) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
> ```objc
>       Eigen::VectorBlock<VectorX<T>>* output) const {
>     unused(context, input, state);
>     DRAKE_THROW_UNLESS(output->size() == 0);
> ```

We're missing a unit test that this fail-safe is effective.

---

*[drake/systems/framework/test/vector_system_test.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1k6MCgA6fKjeA_m_:-KoI1k6MCgA6fKjeA_ma:b1x50o8) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L21)):*
> ```cpp
> // beyond what the constructor does by default.  For all of the Siso overrides,
> // a call count is retained for tests to examine.
> class TestSisoSystem : public VectorSystem<double> {
> ```

_Meta_ This file still has a lot of confusing use of "SISO" acronym.

---

*[drake/systems/framework/test/vector_system_test.cc, line 330 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1wO9iMPpkUer_kBf:-KoI1wO9iMPpkUer_kBg:b-rks4bj) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L330)):*
> ```cpp
>   NoInputContinuousTimeSystem()
>       : VectorSystem<double>(0, 1) {  // Zero inputs, one output.
>     this->DeclareContinuousState(1);  // One state variable.
> ```

FYI _Meta_ I'm not sure these comments are adding much value, in unit test code.

---

*[drake/systems/framework/test/vector_system_test.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI2xtPw2YKgoVvU51s:-KoI2xtPw2YKgoVvU51t:be8qn1g) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L371)):*
> ```cpp
>  public:
>   NoInputNoOutputDiscreteTimeSystem()
>       : VectorSystem<double>(0, 0) {  // Zero inputs, one output.
> ```

"One output" seems like a paste-o.

---

*[drake/systems/framework/test/vector_system_test.cc, line 382 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI3HizTj5HlDI7h_dp:-KoI3HizTj5HlDI7h_dq:b-91p0z8) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L382)):*
> ```cpp
>       const Eigen::VectorBlock<const Eigen::VectorXd>& u,
>       const Eigen::VectorBlock<const Eigen::VectorXd>& x,
>       Eigen::VectorBlock<Eigen::VectorXd>* xn) const {
> ```

Using `xn` the a name for `x[n+1]` is misleading.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-:-KoI7WlsIWL8rOKhbyCx:b-te4kzy)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
addressed review comments.  (will push as soon as the tests finish running locally)

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[drake/BUILD, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1CSgcXR1IaUMqAhJ:-KogI1i0PwTgHDKspN_B:b-biosa5) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/BUILD#L323)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The prior sort-order was correct.
</blockquote></details>

Done.  (any hope of teaching clion's bazel formatter to do the right thing?)

---

*[drake/examples/simple_continuous_time_system.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHuJcEM-pwFqz_tsnB:-KogIU0dtcleTCHLxS3f:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_continuous_time_system.cc#L18)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider a comment on each line:
```
(0,    // Zero inputs.
 1) {  // One output.
```
</blockquote></details>

Done.

---

*[drake/examples/simple_continuous_time_system.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI0jA5kv7xzts7Vt_-:-KogIa_x7t6AHF0mbl7q:b-tdiyun) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_continuous_time_system.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI _Meta_ This override does not use the superclass's names for the parameters; usually this is a defect, though perhaps there is some reason why they should differ in this case?  Should the base class be changed to use these short names instead?
</blockquote></details>

decided it's ok with the long names.  but I have introduced/proposed a change from "discrete_updates" to "next_state", which I think is much more clear.  The term discrete adds no value here, but many characters.  lmk.

---

*[drake/examples/simple_discrete_time_system.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI0SSHD07BQxZTwlOH:-KogJIX2fra_kIpOEFM9:b-qssebl) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/examples/simple_discrete_time_system.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

... and in particular (re: renaming base class's paramter)s , `xn` seems like a really bad name, since this is `x[n+1]` not `x[n]` (which is named `x` above).
</blockquote></details>

fixed.

---

*[drake/systems/framework/vector_system.h, line 52 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHnn_tQcMncGrqwXMz:-KogJw-uFCZfRH4goZ1x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This use of "SISO" is now confusing, because the term is no longer defined above.
</blockquote></details>

Done.

---

*[drake/systems/framework/vector_system.h, line 104 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHoP2_2LwWzJnpHfZ_:-KpZiLkOxhrSlRjKUOcD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L104)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The lifetime of the temporary `VectorX<T>(0)` does not extend past this statement, but `input_block` continues to alias (and use) it.  (It should turn into a named local variable.)

Ditto for the identical code below.
</blockquote></details>

Done.

---

*[drake/systems/framework/vector_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHowjUeIJXIyY58l7z:-KpZiyGwy-2lvhMbl1p6:b-kwtmqc) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This method will only ever be called when there was an output port declared; thus, this guard is misleading.  (The best fix is probably to turn this into an assertion.)

Also, the vector size doesn't seems like the correct thing to check -- its the port count that matters, not the port width, I think?
</blockquote></details>

i agree, but was trying to follow the model above (as in the discrete updates).  i've changed it to check for num_output ports, but left it as a silent fail (like discrete updates), even though I think it should only happen if outputs were declared.

---

*[drake/systems/framework/vector_system.h, line 173 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHpsB1DtbylkNa3CjP:-KpZjKj7AXnzi-7wfPAh:b58ki5p) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L173)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This is the third time this (tricky) bit of code appears.  Normally that would be a defect, but in this case the purpose of the class is to deal with this boilerplate, and its not immediately clear that a factoring would be simpler.  But in addressing the defect (above), please consider whether factoring out this code could be better.
</blockquote></details>

i think it's exactly three times (ever).  avoiding boilerplate for this round.

---

*[drake/systems/framework/vector_system.h, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsQMyVOfVKjwht4zW:-KpZja0oorC_96ebD472:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L206)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`@p derivatives` seems like a paste-o.
</blockquote></details>

Done.

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsyf7En9QyfppQKN7:-KpZjn5fitQJpFj7uQSx:b-ar4zs4) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ditto on "the vector size doesn't seems like the correct thing to check -- its the port count that matters, not the port width, I think?".
</blockquote></details>

this time i think checking the actual input argument (the vector) size is appropriate.  and again it matches the use in the discrete updates version

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI7Laq_11NuoKiMC6C:-KpZtL9Gq3BLU7lXRgjN:bdeyrjv) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

We're missing a unit test that this fail-safe is effective.
</blockquote></details>

agreed.  but then we're also missing them for the failsafes that I copied this from?  what's the best way to write them?  calling DoCalcVectorOutput directly from vector_system_test (and using EXPECT_NO_THROW)? 

---

*[drake/systems/framework/test/vector_system_test.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1k6MCgA6fKjeA_m_:-KpZrihRYedOBhtnW21D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L21)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ This file still has a lot of confusing use of "SISO" acronym.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI2xtPw2YKgoVvU51s:-KpZsRW5owuCePp3fH1F:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L371)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

"One output" seems like a paste-o.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 382 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI3HizTj5HlDI7h_dp:-KpZsGAiJKpwlYPKwGhP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L382)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Using `xn` the a name for `x[n+1]` is misleading.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-:-KpZvgKxZ8PlYIRiNVA7:bskoydy)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 25 files at r1, 6 of 8 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/BUILD, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1CSgcXR1IaUMqAhJ:-Kp_9GCl9z3UxOli0T8r:b-27xyz6) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/BUILD#L323)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.  (any hope of teaching clion's bazel formatter to do the right thing?)
</blockquote></details>

https://github.com/bazelbuild/intellij/issues/73 claims that it is fixed as of May 2017 -- has there been a CLion plugin update published since then?

---

*[drake/systems/framework/vector_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHowjUeIJXIyY58l7z:-Kp_FD1miL6OGVllOnmb:brcyai1) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L164)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i agree, but was trying to follow the model above (as in the discrete updates).  i've changed it to check for num_output ports, but left it as a silent fail (like discrete updates), even though I think it should only happen if outputs were declared.
</blockquote></details>

The difference is that to my understanding `DoCalcDiscreteVariableUpdates` can be called anytime (there is a call graph from user code that reaches it), but due to the code within this file, `CalcVectorOutput` has path from user code to reach it only if there is an output port -- in other words, we've established a class invariant that its impossible to call this method when there are no ports.  Thus, I still believe that an if-check is misleading.

One way to know for sure is to try to write a unit test that exercises the `return` statement -- if you can do so, then I'm wrong and there is no defect (and we gained an important test!); if you can't, then the code is unreachable and should be removed (turned into an assertion).

---

*[drake/systems/framework/vector_system.h, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsQMyVOfVKjwht4zW:-Kp_Dv8SpcTKh4IpX3-l:b-cryxk4) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L206)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Typo `outputs` here vs the parameter name `output`.

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsyf7En9QyfppQKN7:-Kp_GABSozZT96lhkE13:b-xd3vm6) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

this time i think checking the actual input argument (the vector) size is appropriate.  and again it matches the use in the discrete updates version
</blockquote></details>

You're right.

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI7Laq_11NuoKiMC6C:-Kp_GvBoDEuDIUizgIy3:b662vvv) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
> but then we're also missing them for the failsafes that I copied this from?

Indeed.  Please either add that test too, or file an issue to me to add it after this PR.

> what's the best way to write them?

Construct a testing system¹ that passes a non-zero `output_size` to the `VectorSystem` constructor, but does not override this helper.  Do an `EXPECT_THROW` while doing `dut.CalcOutput()` and it should hit this `throw`.

¹ Since `TestVectorSystem` overrides this method currently, we'll either need to declare a different VectorSystem subclass for testing, or else alter `TestVectorSystem`s override to optionally call back into the superclass as if it were not overridden.

---

*[drake/systems/framework/vector_system.h, line 220 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_INP8E-2rg1UK2OSP:-Kp_INP8E-2rg1UK2OSQ:bp1hb24) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/vector_system.h#L220)):*
> ```objc
>   }
> 
>   /// Provides a convenience method for VectorSystem subclasses.  This
> ```

FYI The `%` to disable the self-hyperlink was intentional here, and below.

---

*[drake/systems/framework/test/vector_system_test.cc, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_En0EdRg11B3s8kiG:-Kp_En0EdRg11B3s8kiH:bta4vkz) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L67)):*
> ```cpp
>   }
> 
>   // Vector override.
> ```

FYI `VectorSystem` would be clearer.

---

*[drake/systems/framework/test/vector_system_test.cc, line 168 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_EsOKhvuDBU4YIAAj:-Kp_EsOLKkvnR_lvPTMt:bta4vkz) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L168)):*
> ```cpp
> }
> 
> // Subclasses that violate the Vector premise fail-fast.
> ```

FYI `VectorSystem` would be clearer.

---

*[drake/systems/framework/test/vector_system_test.cc, line 186 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_Ev0pGImEjb3b_ug8:-Kp_Ev0qe7Vzh_bJM0c4:bta4vkz) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L186)):*
> ```cpp
>   { // Abstract state.
>     // TODO(jwnimmer-tri) Conceptually, allowing abstract state could be
>     // consistent with the Vector contract.  However, if we widened the Vector
> ```

FYI `VectorSystem` would be clearer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/BUILD, line 323 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1CSgcXR1IaUMqAhJ:-KpfnW01y7l4RZ4YAtIy:bjreewb) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/BUILD#L323)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

https://github.com/bazelbuild/intellij/issues/73 claims that it is fixed as of May 2017 -- has there been a CLion plugin update published since then?
</blockquote></details>

since we're behind in the clion versions (according to my watching of slack, and the  current sphinx docs, i'm still on 2016), it doesn't look like there are any newer versions of the buildifier available.

---

*[drake/systems/framework/vector_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHowjUeIJXIyY58l7z:-KpfppZR_yo8Mc0oy--i:b-rzms06) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The difference is that to my understanding `DoCalcDiscreteVariableUpdates` can be called anytime (there is a call graph from user code that reaches it), but due to the code within this file, `CalcVectorOutput` has path from user code to reach it only if there is an output port -- in other words, we've established a class invariant that its impossible to call this method when there are no ports.  Thus, I still believe that an if-check is misleading.

One way to know for sure is to try to write a unit test that exercises the `return` statement -- if you can do so, then I'm wrong and there is no defect (and we gained an important test!); if you can't, then the code is unreachable and should be removed (turned into an assertion).
</blockquote></details>

I agree , and have changed to assert.

---

*[drake/systems/framework/vector_system.h, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoHsQMyVOfVKjwht4zW:-Kpfq7oGoV4Xb6O4712L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L206)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Typo `outputs` here vs the parameter name `output`.
</blockquote></details>

Done.

---

*[drake/systems/framework/vector_system.h, line 214 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI7Laq_11NuoKiMC6C:-KpfuT3lUkjF96csoFEQ:b-wrzkb8) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/vector_system.h#L214)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> but then we're also missing them for the failsafes that I copied this from?

Indeed.  Please either add that test too, or file an issue to me to add it after this PR.

> what's the best way to write them?

Construct a testing system¹ that passes a non-zero `output_size` to the `VectorSystem` constructor, but does not override this helper.  Do an `EXPECT_THROW` while doing `dut.CalcOutput()` and it should hit this `throw`.

¹ Since `TestVectorSystem` overrides this method currently, we'll either need to declare a different VectorSystem subclass for testing, or else alter `TestVectorSystem`s override to optionally call back into the superclass as if it were not overridden.
</blockquote></details>

added them all.

---

*[drake/systems/framework/vector_system.h, line 220 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_INP8E-2rg1UK2OSP:-KpfqlsbMhoPK2DkHBuR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/vector_system.h#L220)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI The `%` to disable the self-hyperlink was intentional here, and below.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 330 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KoI1wO9iMPpkUer_kBf:-KpfrKKeYJ5m63aB7A5B:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9496043c2301fb57f72fd5980d9b1209154fbe3b/drake/systems/framework/test/vector_system_test.cc#L330)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI _Meta_ I'm not sure these comments are adding much value, in unit test code.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_En0EdRg11B3s8kiG:-Kpfr2qwpmSWe_4ZAJiq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L67)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `VectorSystem` would be clearer.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 168 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_EsOKhvuDBU4YIAAj:-Kpfr3JyR_sY9FCY8EQS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L168)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `VectorSystem` would be clearer.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 186 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kp_Ev0pGImEjb3b_ug8:-KpfrA1JmXUtET_Tp6bo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9ee5ac252e5636723406b40d35baa75029b9768a/drake/systems/framework/test/vector_system_test.cc#L186)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `VectorSystem` would be clearer.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 25 files at r1, 8 of 9 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
:lgtm:

+@EricCousineau-TRI for platform review, please.

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, with minor comments. Thanks!

---

Reviewed 14 of 25 files at r1, 7 of 9 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/vector_system.h, line 248 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kporj4Sczr2IpeCH55a:-Kporj4TttbIjfJRCtd_:bjcmemb) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/vector_system.h#L248)):*
> ```objc
>   /// whether discrete state was declared at context-creation time.
>   ///
>   /// By default, this function does nothing if the @p next_state are
> ```

nit Grammar, "next_state is empty"

---

*[drake/systems/framework/test/vector_system_test.cc, line 186 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kpos6GqyzVJSW9Yh1xV:-Kpos6GrznwrJgwo8kcF:b-v0vdup) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/test/vector_system_test.cc#L186)):*
> ```cpp
>     // TODO(jwnimmer-tri) Conceptually, allowing abstract state could be
>     // consistent with the VectorSystem contract.  However, if we widened the
>     // Vector
> ```

Should `Vector` be `VectorSystem` here?

---

*[drake/systems/framework/test/vector_system_test.cc, line 353 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KposgEeCCLWFILyAzPD:-KposgEfyggYrk8xn5qg:biruoj0) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/test/vector_system_test.cc#L353)):*
> ```cpp
> 
> // Derivatives and Output methods still work when input size is zero.
> TEST_F(VectorSystemTest, NoInputSystemTest) {
> ```

BTW Should this test be renamed to `NoInputContinuousTimeSystemTest`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-:-KpotsAwyinSqOKCyXfV:bz8i4c2)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/vector_system.h, line 248 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kporj4Sczr2IpeCH55a:-KppS-saMOR6QboAc-Hl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/vector_system.h#L248)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit Grammar, "next_state is empty"
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 186 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-Kpos6GqyzVJSW9Yh1xV:-KppSOAGFCHMGRsvJHKW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/test/vector_system_test.cc#L186)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Should `Vector` be `VectorSystem` here?
</blockquote></details>

Done.

---

*[drake/systems/framework/test/vector_system_test.cc, line 353 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6501#-KposgEeCCLWFILyAzPD:-KppSYXg8N2a1L9zBPrI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/46907246d35d034f574c2487b343029194d5fc4f/drake/systems/framework/test/vector_system_test.cc#L353)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Should this test be renamed to `NoInputContinuousTimeSystemTest`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 25 files at r1, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6501)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Some of the review fixups got rebased into the final commit (examples) instead of first commit (SISO rename).  As such, I'm not going to merge this myself.  I'll let @RussTedrake decide if he wants to refactor the commits, squash-merge, or merge anyway.
